dstep/translator/Context.d

```diff
@@ -198,9 +219,72 @@ bool shouldBeAnonymous(Context context, Cursor cursor)
 }
 
 /**
-  * Returns true, if cursor is in the global scope.
-  */
+ * Returns true, if cursor is in the global scope.
+ */
 bool isGlobal(Cursor cursor)
 {
     return cursor.semanticParent.kind == CXCursorKind.CXCursor_TranslationUnit;
 }
+
+/**
+ * The collectTypeNames function scans the whole AST of the translation unit and produces
+ * a set of the type names in global scope.
+ *
+ * The type names are required for the parsing of C code (e.g. macro definition bodies),
+ * as C grammar isn't context free.
+ */
+
+bool[string] collectTypeNames(TranslationUnit translUnit)
```

Comment:
> It seems that (un)signed can be specified in any position, before, after and in between (in the case of long long)
> short int, long int, long double are also valid types
> const, volatile and _Atomic (C11) can be specified for types as well
> __complex__ and _Complex in any combination of the floating point types are valid as well. The > floating point type does not need to be specified
> _Bool and bool (if <stdbool.h> is included) are valid types in C99. I tried compiling code containing these types with -std=c89 -Wall and got no errors or warnings
> When a type consist of multiple words, i.e. long int, any order of the words are valid
> What about the spaces. If the source code specifies "long long" (seems like multiple spaces are removed by GitHub), will the spelling contain that as well or does Clang remove the extra spaces?

> What about unions and enums?

Well, I need to write parsing of the types. It will not be hard, but there is another problem. Even if I parse the the types, then I need to translate them. But, there are routines for translating types in dstep already, most in `dstep/translator/Type.d`. But they only accept information in the form of `Type` and `Cursor` objects. It would be stupid to have separate code path to translate types in defines.

This is related to what I said in #65 (desing problem). One way would be to make all the functions templates, but it doesn't seem doable in practice.

I propose to change a `Type` type from struct to class and then create two new classes that inherits from `Type`: `CLangType` and let say `UserType`. `CLangType` would talk to cx and access clang interface directly, `UserType` would have a constructor allowing an user to create it and fill with data directly.

Another way would be to create a `TypeImpl` struct and store both pointer to `TypeImpl` and `cx` inside `Type` and in every method check which one to use with if.

```
struct TypeImpl
{
    string spelling;
    bool isConst;
    etc...
};

struct Type
{
    mixin CX;
    TypeImpl* impl;

    string spelling() 
    {
        return impl ? impl.spelling : /* get spelling from cx */;
    }
}

```

Another way is to fill `Type` with all required fields and copy them on creation.

```
struct Type
{
    string spelling;
    bool isConst;
    bool isVolatile;
    Type* canonicalType;
    etc.

    Type(void* returned_from_clang)
    {
        // get data from libclang and fill fields
    }
}
```

What do you think about it?